src/main/java/org/truffleruby/core/regexp/InterpolatedRegexpNode.java

```diff
@@ -9,71 +9,115 @@
+            return RegexpBuilderNodeGen.create(options, null);
         }
 
-        final RopeBuilder preprocessed = ClassicRegexp.preprocessDRegexp(getContext(), strings, options);
+        public RegexpBuilderNode(RegexpOptions options) {
+            this.options = options;
+        }
+
+        public abstract DynamicObject execute(Rope[] parts);
 
-        final DynamicObject regexp = RegexpNodes.createRubyRegexp(getContext(), this, coreLibrary().getRegexpFactory(),
-                RopeOperations.ropeFromByteList(preprocessed), options);
+        @Specialization(guards = { "ropesMatch(cachedParts, parts)" }, limit = "getCacheLimit()")
+        public DynamicObject executeFast(Rope[] parts,
+                @Cached(value = "parts", dimensions = 1) Rope[] cachedParts,
+                @Cached("createRegexp(cachedParts)") DynamicObject regexp) {
+            return regexp;
+        }
 
-        if (options.isEncodingNone()) {
-            final Rope source = Layouts.REGEXP.getSource(regexp);
+        @Specialization(replaces = "executeFast")
+        public DynamicObject executeslow(Rope[] parts) {
+            Log.notOptimizedOnce(Log.UNSTABLE_INTERPOLATED_REGEXP);
+            return createRegexp(parts);
+        }
 
-            if (!BodyTranslator.all7Bit(preprocessed.getBytes())) {
-                Layouts.REGEXP.setSource(regexp, RopeOperations.withEncodingVerySlow(source, ASCIIEncoding.INSTANCE));
-            } else {
-                Layouts.REGEXP.setSource(regexp, RopeOperations.withEncodingVerySlow(source, USASCIIEncoding.INSTANCE));
+        @ExplodeLoop
+        protected boolean ropesMatch(Rope[] a, Rope[] b) {
+            for (int i = 0; i < a.length; i++) {
+                if (!ropesEqualNode.execute(a[i], b[i])) {
+                    return false;
+                }
             }
+            return true;
         }
 
-        return regexp;
-    }
+        @TruffleBoundary
+        protected DynamicObject createRegexp(Rope[] strings) {
+            final RopeBuilder preprocessed = ClassicRegexp.preprocessDRegexp(getContext(), strings, options);
 
-    @ExplodeLoop
-    protected DynamicObject[] executeChildren(VirtualFrame frame) {
-        DynamicObject[] values = new DynamicObject[children.length];
-        for (int i = 0; i < children.length; i++) {
-            final Object value = children[i].execute(frame);
-            values[i] = (DynamicObject) toS.call(frame, value, "to_s");
+            final DynamicObject regexp = RegexpNodes.createRubyRegexp(getContext(), this, coreLibrary().getRegexpFactory(),
+                    RopeOperations.ropeFromByteList(preprocessed), options);
+
+            if (options.isEncodingNone()) {
+                final Rope source = Layouts.REGEXP.getSource(regexp);
+
+                if (!BodyTranslator.all7Bit(preprocessed.getBytes())) {
+                    Layouts.REGEXP.setSource(regexp, RopeOperations.withEncodingVerySlow(source, ASCIIEncoding.INSTANCE));
```

Comment:
I'm not entirely clear on the relationship between `preprocessed` and `source`. You might need to call `RopeOperations.withEncodingVerySlow(source, ASCIIEncoding.INSTANCE, CodeRange.CR_UNKNOWN)` instead here.

Reply:
Nevermind. I see this code was there already. But now that I'm looking at it it does seem a bit suspect. Since you've been looking at this code in general, if you can provide any insight into the open question, that'd be helpful.

Reply:
I'm not sure what's going on with that. It goes back to the start of branching off JRuby as far as I can see. Maybe @chrisseaton could shed some light on that. Commenting out those lines doesn't cause any core specs to fail, or any of the MRI tests.

Reply:
I guess you would need to check the regexp's source encoding to verify. But the way this section is written, it looks like `preprocessed` and `source` are tightly related somehow. Otherwise, I don't know how you could trust one to dictate the encoding (and subsequent code range) of the other.

Reply:
Ah, right. `source` is closely related to `preprocessed` but the act of compiling the regexp may force an encoding change, so `source` is set from the `userObject` on the regexp.

Reply:
So what this boils down to is should the `withEncodingVerySlow` call be made with the 3 argument form where the code range is explicitly passed? And should the first branch call with `CR_UNKNOWN`? But, again, this code was already there so I'm not looking to block the merge on this. It just got highlighted and I thought it might be worth checking it out.